### PR TITLE
fix: revert esm changes

### LINKED ISSUE #195: BUG: ESM imports not working
Version 6.0.0 package.json is missing the "main" key which prevents the library from being imported. It was in package.json before, I assume it's being missing is just a typo. 

Nothing else to add, pretty straightforward to fix.

### PR DESCRIPTION
**Pull request by Nebulis**
closes #195 

This PR reverts all the changes to make OA an ESM-only library. The reasons for the revert are detailed below.

## Disclaimer

Once OA v6.0.0 was released, I tried to use it in another library. I faced many problems, specifically with Jest, and decided to dig more and check whether there were issues with other tools.

I noticed that the library built didn't work directly (without specific configuration) with the following tools:
- node
- ts-node
- jest
- ts-jest
- webpack

#### Path
When trying to use the library in a node project, the following error happened:
```txt
Directory import '/home/nebulis/IdeaProjects/test-esm/node_modules/@govtechsg/open-attestation/dist/shared/validate' is not supported resolving ES modules imported from /home/nebulis/IdeaProjects/test-esm/node_modules/@govtechsg/open-attestation/dist/index.js
```

The reason is simple: we didn't use full path to import files. We relied on common js behavior, which doesn't require the extension (.js) and which automatically loads index files if it exists.

To fix there is a [node flag](https://nodejs.org/api/esm.html#esm_customizing_esm_specifier_resolution_algorithm):

```sh
❯ node --experimental-specifier-resolution=node src/index.js
```

#### CommonJS
Once the path is fixed for node, a new error happened:

```txt
SyntaxError: Named export 'identity' not found. The requested module 'lodash' is a CommonJS module, which may not support all module.exports as named exports.
```

There is no way to fix this with a flag. We must fix our library and changed all imports that are from CJS libraries. To manage to test our library with the rest of the tools, I fixed manually, inside the node_modules folder. Once fixed, the library was usable with node.

Additional readings:
- https://simonplend.com/node-js-now-supports-named-imports-from-commonjs-modules-but-what-does-that-mean/
- https://redfin.engineering/node-modules-at-war-why-commonjs-and-es-modules-cant-get-along-9617135eeca1

### Jest

To have Jest working, the first thing is to follow [the documentation](https://jestjs.io/docs/ecmascript-modules):
  - add `transform {}` (useless)
  - add `NODE_OPTIONS=--experimental-vm-modules`
  
And additionally we need a custom resolver (answer found somewhere tin the Jest issues:
  - add custom resolver => `"resolver": "./export_maps_resolver.cjs"`
```js
// export_maps_resolver.cjs
const resolver = require("enhanced-resolve").create.sync({
  conditionNames: ["require", "node", "default", "import"],
  extensions: [".js", ".json", ".node", ".ts"],
});

module.exports = function (request, options) {
  if (request === "crypto") {
    return "crypto";
  }
  return resolver(options.basedir, request);
};
```

Additional readings:
- https://github.com/facebook/jest/issues/9771
- https://github.com/facebook/jest/issues/9771#issuecomment-841624042
- https://github.com/facebook/jest/issues/9430

### ts-kest

To have ts-jest working, the first thing is to follow [the documentation](https://kulshekhar.github.io/ts-jest/docs/guides/esm-support), in addition to the steps used for Jest above:
```json
{
  // jest config
  "preset": "ts-jest",
  "testEnvironment": "node",
  "extensionsToTreatAsEsm": [".ts"],
  "globals": {
    "ts-jest": {
      "useESM": true
    }
  }
}
```

### ts-node
When using ts-node, the following error happened:
```txt
Unknown file extension ".ts"
```

ts-node doesn't support ESM for the moment. The progress can be followed in https://github.com/TypeStrong/ts-node/issues/1007. There is an experimental loader that can be used by following the direction in the issue:
- add `"module": "ESNext"` into tsconfig.json
- add `"type": "module"` into package.json
- npm install ts-node
- node --experimental-specifier-resolution=node --loader ts-node/esm src/index.js

It's important to use ts-node as a loader.

## Webpack

Webpack didn't work for the same reason as node. The imports are not valid. There is a flag to set, to force webpack to infer the path:

```js
// webpack.config.js
{
  test: /\.m?js/,
  resolve: {
    fullySpecified: false
  }
}
```

Additional reading:
- https://github.com/webpack/webpack/issues/11467#issuecomment-691873586

### Issues

Overall there are 2 issues

1. Our import paths are wrong. All our paths should directly target the files, without any inference required.
2. We must import differently CJS modules:
- lodash
- hs-sha3
- flatley

#### Path
We could expect typescript to help us, but no. Like for [custom paths](https://github.com/microsoft/TypeScript/issues/10866), maintainers decided that it's not their job. So we are left alone. You can find a discussion on the topic below:
- https://github.com/microsoft/TypeScript/issues/16577

There are tools to help:
- https://github.com/Zoltu/typescript-transformer-append-js-extension

I tried to find a lint rule to help, but it looks like ESM and Typescript don't move together in the same direction:
- https://github.com/import-js/eslint-plugin-import/issues/2170

If you read carefully the discussion, you will find out that you can import `.js` files, inside typescript files, even if the file you import is a typescript file (and not a javascript file). Heresy? :)

This solution works, however, I faced issues with ts-jest. Some people [made it work](https://github.com/kulshekhar/ts-jest/discussions/2475), but at this point, I was over with this, and decided to not dig more

#### CJS modules

The fix is quite simple, but I tried to find a way to trigger an error on build or on lint. Unsuccessfully.

## Conclusion

With all this information, we decided to revert back to our old process. ESM and TS don't play well together and we may reconsider in the future if there are improvements with the tooling. The goal must be to help library users and developers but right now, it will just add a burden for both.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -2,19 +2,18 @@
...
   "scripts": {
     "benchmark:qr-code": "ts-node --transpile-only benchmarks/qr-code",
-    "build": "npm run build:clean && npm run build:esm",
-    "build:esm": "tsc --outDir dist --declaration --project ./tsconfig.prod.json",
-    "build:clean": "rm -rf dist/",
+    "build": "npm run clean && npm run build:cjs && npm run build:esm && npm run build:umd && npm run build:type",
+    "build:cjs": "tsc --module commonjs --outDir dist/cjs --project ./tsconfig.prod.json",
+    "build:esm": "tsc --module es2015 --outDir dist/esm --project ./tsconfig.prod.json",
+    "build:type": "tsc -d --emitDeclarationOnly --outDir dist/types --project ./tsconfig.prod.json",
+    "build:umd": "rollup -c",
```
Comment (eugbyte) [inline package.json:15]: since rollup is building from `./dist/esm/index.js`, should we do
```
"build:umd": "npm run build:cjs && rollup -c",
```

**Review by eugbyte — APPROVED** [review]

File: package.json (hunk 1 above)
Comment (eugbyte) [reply package.json:15]: need to update the circle i to install rollup

Comment (Nebulis) [reply package.json:15]: hum no I will keep it as it was before

Comment (Nebulis) [reply package.json:15]: for rollup I forgot to put it back. Fixed

**Pull request merged**